system/lib/pthread/emscripten_futex_wait.c

```diff
@@ -143,7 +145,25 @@ int emscripten_futex_wait(volatile void *addr, uint32_t val, double max_wait_ms)
   if (max_wait_ms != INFINITY) {
     max_wait_ns = (int64_t)(max_wait_ms*1000*1000);
   }
+#ifdef __PIC__
+  // After the main thread queues dlopen events, it checks if the target threads
+  // are sleeping.
+  // If `sleeping` is set then the main thread knows that event will be
+  // processed after the sleep (before any other user code).  In this case the
+  // main thread does not wait for any kind of response form the thread.
+  // If `sleeping` is not set then we know we can expect a responce from the
+  // thread, either from the call here directly after settings sleeping = 1, or
+  // from another callside (e.g.  the one in scripten_yield).
+  __pthread_self()->sleeping = 1;
+  _emscripten_process_dlopen_queue();
```

Comment:
Why do it both before and after? Isn't after enough?

Reply:
Executing even an empty queue requires taking two locks, so it's not quite as cheap as a single atomic operation.

Reply:
> Executing even an empty queue requires taking two locks, so it's not quite as cheap as a single atomic operation.

Is there some way to early exit faster than that when the q is empty?   I don't think it matter a huge amount for this PR because we have already decided to actually sleep...

Reply:
No, because `em_proxying_queue` is actually a list of thread-specific `em_task_queues`, it requires taking a lock just to find the right task queue and another lock to check if the task queue is empty. We could do [double-checked locking](https://en.wikipedia.org/wiki/Double-checked_locking) to avoid taking the second lock when the queue is empty or even investigate lock-free queues, but I wouldn't want to try blindly optimizing this.

Reply:
I guess for system queues like the dlopen queue, we could extract the `em_task_queue` and store it in TLS to avoid the first lock as well.

Reply:
> Another way of thinking about it, we need to process the queue here in case there was already message in it before we set sleeping = 1.

I don't see the problem yet, sorry. If there was already a message in it, the message can wait there until the place you currently have the second call to process the queue. Maybe I'm not seeing how processing the queue interacts with setting `sleeping = 1` somehow?

What I am suggesting is that you remove the first of the two calls to process the queue, which leaves the later one, the second, in place. I think that is enough, because

1. Leaving the later one in place means that we will process all the events. (Maybe a little later.)
2. The other thread does not notice whether we've processed all the events earlier, or not.

But I'm probably missing something here...

Reply:
The `sleeping = 1` check determines weather the main thread will block/wait until the event is processed.

In the normal case the main thread blocks until the event is processed.  The case we want to avoid is when a thread goes to sleep (potentially) forever while the main thread waiting for a response.   Coming into this function there could already be a waited-on even in queue, and we can't go to sleep with a waited-on event in the queue, otherwise we deadlock.   

Basically, we have have to process the queue before we go to sleep.    Its not enough to just to it after, because in the case we block forever that time might never come.